File: private-ca/utils/issuer-alt-name.py
#!/usr/bin/env python3

from cryptography import x509
import base64
import argparse

parser = argparse.ArgumentParser()
parser.add_argument("-e", "--email", type=str)
parser.add_argument("-u", "--uri", type=str)

args = parser.parse_args()

if not (args.email or args.uri):
    parser.error('At least one of --email or --uri must be specified')

general_names = []

if args.email:
    general_names.append(x509.RFC822Name(args.email))

if args.uri:
    general_names.append(x509.UniformResourceIdentifier(args.uri))

extension = x509.IssuerAlternativeName(general_names)

print(base64.b64encode(extension.public_bytes()))

File: platform-ca/utils/issuer-alt-name.ts
import { AsnSerializer } from "@peculiar/asn1-schema";
import { GeneralName, GeneralNames, IssueAlternativeName } from "@peculiar/asn1-x509";
import { Buffer } from "buffer";
import { Command } from "commander";

const program = new Command();

function issuerAltName(program: Command) {
  program
    .name("issuer-alt-name")
    .description("CLI tool to generate a base64-encoded ASN.1 representation of an Issuer Alternative Name for an X.509v3 certificate extension")
    .version("0.0.1")
    .option("-e, --email <email>", "Email address to include in the Issuer Alternative Name")
    .option("-u, --uri <uri>", "URI to include in the Issuer Alternative Name")
    .action((opts, command) => {
      if (!opts.email && !opts.uri) {
        command.error("At least one of email address and URI must be supplied");
      }
    });

  program.parse();

  const generalNames: GeneralNames = [];

  if (program.opts().email) {
    generalNames.push(
      new GeneralName({
        rfc822Name: program.opts().email,
      }),
    );
  }

  if (program.opts().uri) {
    generalNames.push(
      new GeneralName({
        uniformResourceIdentifier: program.opts().uri,
      }),
    );
  }

  try {
    const asn1Structure = new IssueAlternativeName(generalNames);
    const serializedAsn1 = AsnSerializer.serialize(asn1Structure);
    const base64EncodedSerializedAsn1 = Buffer.from(serializedAsn1).toString("base64");
    console.log(base64EncodedSerializedAsn1);
  } catch (e) {
    console.error("An error occurred generating the Issuer Alternative Name:", e);
    process.exit(1);
  }
}

issuerAltName(program);
